BUG: invalid type annotation for equal operator of Series

- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df = pd.DataFrame(
	{'kind' : ['a', 'b', 'a'], 
	 'value' : [1, 2, 3]})

res: pd.Series = df['kind'] == 'a'
print(type(res))
```

#### Problem description

`mypy` returns the following error:

```
$ mypy test.py 
test.py:7: error: Incompatible types in assignment (expression has type "bool", variable has type
"Series")
    res: pd.Series = df['kind'] == 'a'
                     ^
Found 1 error in 1 file (checked 1 source file)
$ mypy --version
mypy 0.812
```

But actual type of the result at runtime is `Series`:

```
$ python test.py 
<class 'pandas.core.series.Series'>
```

#### Expected Output

No compile errors.

#### Output of ``pd.show_versions()``

<details>

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : f2c8480af2f25efdbd803218b9d87980f416563e
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-70-generic
Version          : #78-Ubuntu SMP Fri Mar 19 13:29:52 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.3
numpy            : 1.20.2
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 44.0.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None
```

</details>

--- a/doc/source/whatsnew/v3.1.0.rst
+++ b/doc/source/whatsnew/v3.1.0.rst
@@ -37,6 +37,7 @@ Other enhancements
 - Display formatting for float sequences in DataFrame cells now respects the ``display.precision`` option (:issue:`60503`).
 - Improved the precision of float parsing in :func:`read_csv` (:issue:`64395`)
 - Improved the string ``repr`` of :class:`pd.core.arrays.SparseArray` (:issue:`64547`)
+- Improved type inference of comparison and arithmetic operators on :class:`Series` and :class:`DataFrame` for static type checkers (e.g. ``ser == "a"`` is now inferred as :class:`Series` instead of ``Any``) (:issue:`40762`)
 - MSVC is no longer required to build on Windows, and build errors when using the MinGW compiler have been fixed (:issue:`63160`)
 
 .. ---------------------------------------------------------------------------

--- a/pandas/core/arraylike.py
+++ b/pandas/core/arraylike.py
@@ -8,7 +8,10 @@ Methods that can be shared by many array-like classes or subclasses:
 from __future__ import annotations
 
 import operator
-from typing import Any
+from typing import (
+    Any,
+    Self,
+)
 
 import numpy as np
 
@@ -38,27 +41,27 @@ class OpsMixin:
         return NotImplemented
 
     @unpack_zerodim_and_defer("__eq__")
-    def __eq__(self, other):
+    def __eq__(self, other) -> Self:  # type: ignore[override]
         return self._cmp_method(other, operator.eq)
 
     @unpack_zerodim_and_defer("__ne__")
-    def __ne__(self, other):
+    def __ne__(self, other) -> Self:  # type: ignore[override]
         return self._cmp_method(other, operator.ne)
 
     @unpack_zerodim_and_defer("__lt__")
-    def __lt__(self, other):
+    def __lt__(self, other) -> Self:
         return self._cmp_method(other, operator.lt)
 
     @unpack_zerodim_and_defer("__le__")
-    def __le__(self, other):
+    def __le__(self, other) -> Self:
         return self._cmp_method(other, operator.le)
 
     @unpack_zerodim_and_defer("__gt__")
-    def __gt__(self, other):
+    def __gt__(self, other) -> Self:
         return self._cmp_method(other, operator.gt)
 
     @unpack_zerodim_and_defer("__ge__")
-    def __ge__(self, other):
+    def __ge__(self, other) -> Self:
         return self._cmp_method(other, operator.ge)
 
     # -------------------------------------------------------------
@@ -68,27 +71,27 @@ class OpsMixin:
         return NotImplemented
 
     @unpack_zerodim_and_defer("__and__")
-    def __and__(self, other):
+    def __and__(self, other) -> Self:
         return self._logical_method(other, operator.and_)
 
     @unpack_zerodim_and_defer("__rand__")
-    def __rand__(self, other):
+    def __rand__(self, other) -> Self:
         return self._logical_method(other, roperator.rand_)
 
     @unpack_zerodim_and_defer("__or__")
-    def __or__(self, other):
+    def __or__(self, other) -> Self:
         return self._logical_method(other, operator.or_)
 
     @unpack_zerodim_and_defer("__ror__")
-    def __ror__(self, other):
+    def __ror__(self, other) -> Self:
         return self._logical_method(other, roperator.ror_)
 
     @unpack_zerodim_and_defer("__xor__")
-    def __xor__(self, other):
+    def __xor__(self, other) -> Self:
         return self._logical_method(other, operator.xor)
 
     @unpack_zerodim_and_defer("__rxor__")
-    def __rxor__(self, other):
+    def __rxor__(self, other) -> Self:
         return self._logical_method(other, roperator.rxor)
 
     # -------------------------------------------------------------
@@ -98,7 +101,7 @@ class OpsMixin:
         return NotImplemented
 
     @unpack_zerodim_and_defer("__add__")
-    def __add__(self, other):
+    def __add__(self, other) -> Self:
         """
         Get Addition of DataFrame and other, column-wise.
 
@@ -190,63 +193,63 @@ class OpsMixin:
         return self._arith_method(other, operator.add)
 
     @unpack_zerodim_and_defer("__radd__")
-    def __radd__(self, other):
+    def __radd__(self, other) -> Self:
         return self._arith_method(other, roperator.radd)
 
     @unpack_zerodim_and_defer("__sub__")
-    def __sub__(self, other):
+    def __sub__(self, other) -> Self:
         return self._arith_method(other, operator.sub)
 
     @unpack_zerodim_and_defer("__rsub__")
-    def __rsub__(self, other):
+    def __rsub__(self, other) -> Self:
         return self._arith_method(other, roperator.rsub)
 
     @unpack_zerodim_and_defer("__mul__")
-    def __mul__(self, other):
+    def __mul__(self, other) -> Self:
         return self._arith_method(other, operator.mul)
 
     @unpack_zerodim_and_defer("__rmul__")
-    def __rmul__(self, other):
+    def __rmul__(self, other) -> Self:
         return self._arith_method(other, roperator.rmul)
 
     @unpack_zerodim_and_defer("__truediv__")
-    def __truediv__(self, other):
+    def __truediv__(self, other) -> Self:
         return self._arith_method(other, operator.truediv)
 
     @unpack_zerodim_and_defer("__rtruediv__")
-    def __rtruediv__(self, other):
+    def __rtruediv__(self, other) -> Self:
         return self._arith_method(other, roperator.rtruediv)
 
     @unpack_zerodim_and_defer("__floordiv__")
-    def __floordiv__(self, other):
+    def __floordiv__(self, other) -> Self:
         return self._arith_method(other, operator.floordiv)
 
     @unpack_zerodim_and_defer("__rfloordiv")
-    def __rfloordiv__(self, other):
+    def __rfloordiv__(self, other) -> Self:
         return self._arith_method(other, roperator.rfloordiv)
 
     @unpack_zerodim_and_defer("__mod__")
-    def __mod__(self, other):
+    def __mod__(self, other) -> Self:
         return self._arith_method(other, operator.mod)
 
     @unpack_zerodim_and_defer("__rmod__")
-    def __rmod__(self, other):
+    def __rmod__(self, other) -> Self:
         return self._arith_method(other, roperator.rmod)
 
     @unpack_zerodim_and_defer("__divmod__")
-    def __divmod__(self, other):
+    def __divmod__(self, other) -> tuple[Self, Self]:
         return self._arith_method(other, divmod)
 
     @unpack_zerodim_and_defer("__rdivmod__")
-    def __rdivmod__(self, other):
+    def __rdivmod__(self, other) -> tuple[Self, Self]:
         return self._arith_method(other, roperator.rdivmod)
 
     @unpack_zerodim_and_defer("__pow__")
-    def __pow__(self, other):
+    def __pow__(self, other) -> Self:
         return self._arith_method(other, operator.pow)
 
     @unpack_zerodim_and_defer("__rpow__")
-    def __rpow__(self, other):
+    def __rpow__(self, other) -> Self:
         return self._arith_method(other, roperator.rpow)
 
 

--- a/pandas/core/generic.py
+++ b/pandas/core/generic.py
@@ -11660,7 +11660,7 @@ class NDFrame(PandasObject, indexing.IndexingMixin):
             # We want to restore the original index
             rs = rs.loc[~rs.index.duplicated()]
             rs = rs.reindex_like(self)
-        return rs.__finalize__(self, method="pct_change")
+        return rs.__finalize__(self, method="pct_change")  # type: ignore[return-value]
 
     @final
     def _logical_func(

--- a/pandas/core/indexes/range.py
+++ b/pandas/core/indexes/range.py
@@ -1364,7 +1364,7 @@ class RangeIndex(Index):
         return type(self)._simple_new(res, name=self._name)
 
     @unpack_zerodim_and_defer("__floordiv__")
-    def __floordiv__(self, other: object) -> Index:
+    def __floordiv__(self, other: object) -> Index:  # type: ignore[override]
         if is_integer(other) and other != 0:
             if len(self) == 0 or (self.start % other == 0 and self.step % other == 0):
                 start = self.start // other
@@ -1377,7 +1377,7 @@ class RangeIndex(Index):
                 new_range = range(start, start + 1, 1)
                 return self._simple_new(new_range, name=self._name)
 
-        return super().__floordiv__(other)  # type: ignore[no-untyped-call]
+        return super().__floordiv__(other)
 
     # --------------------------------------------------------------------
     # Reductions

--- a/pandas/io/stata.py
+++ b/pandas/io/stata.py
@@ -213,7 +213,7 @@ def _stata_elapsed_date_to_datetime_vec(dates: Series, fmt: str) -> Series:
         year = stata_epoch.year + dates // 52
         days = (dates % 52) * 7
         per_y = (year - 1970).array.view("Period[Y]")
-        per_d = per_y.asfreq("D", how="S")
+        per_d = per_y.asfreq("D", how="S")  # type: ignore[union-attr]
         per_d_shifted = per_d + days._values
         per_s = per_d_shifted.asfreq("s", how="S")
         conv_dates_arr = per_s.view("M8[s]")

--- a/pandas/plotting/_matplotlib/core.py
+++ b/pandas/plotting/_matplotlib/core.py
@@ -2146,10 +2146,9 @@ class PiePlot(MPLPlot):
 
     def __init__(self, data: Series | DataFrame, kind=None, **kwargs) -> None:
         data = data.fillna(value=0)
-        lt_zero = data < 0
-        if isinstance(data, ABCDataFrame) and lt_zero.any().any():
+        if isinstance(data, ABCDataFrame) and (data < 0).any().any():
             raise ValueError(f"{self._kind} plot doesn't allow negative values")
-        elif isinstance(data, ABCSeries) and lt_zero.any():
+        elif isinstance(data, ABCSeries) and (data < 0).any():
             raise ValueError(f"{self._kind} plot doesn't allow negative values")
         MPLPlot.__init__(self, data, kind=kind, **kwargs)
 

